tests/pytests/unit/modules/test_yumpkg.py

```diff
@@ -1747,3 +1921,36 @@ def test_call_yum_with_kwargs():
                 python_shell=True,
                 username="Darth Vader",
             )
+
+
+@pytest.mark.skipif(not salt.utils.systemd.booted(), reason="Requires systemd")
+@patch("salt.modules.yumpkg._yum", Mock(return_value="dnf"))
+def test_services_need_restart():
+    """
+    Test that dnf needs-restarting output is parsed and
+    salt.utils.systemd.pid_to_service is called as expected.
+    """
+    expected = ["firewalld", "salt-minion"]
+
+    dnf_mock = Mock(
+        return_value="123 : /usr/bin/firewalld\n456 : /usr/bin/salt-minion\n"
+    )
+    systemd_mock = Mock(side_effect=["firewalld", "salt-minion"])
+    with patch.dict(yumpkg.__salt__, {"cmd.run_stdout": dnf_mock}), patch(
+        "salt.utils.systemd.pid_to_service", systemd_mock
+    ):
+        assert sorted(yumpkg.services_need_restart()) == expected
+        systemd_mock.assert_has_calls([call("123"), call("456")])
+
+
+@patch("salt.modules.yumpkg._yum", Mock(return_value="dnf"))
+def test_services_need_restart_requires_systemd():
+    """Test that yumpkg.services_need_restart raises an error if systemd is unavailable."""
+    with patch("salt.utils.systemd.booted", Mock(return_value=False)):
+        pytest.raises(CommandExecutionError, yumpkg.services_need_restart)
+
+
+@patch("salt.modules.yumpkg._yum", Mock(return_value="yum"))
+def test_services_need_restart_requires_dnf():
+    """Test that yumpkg.services_need_restart raises an error if DNF is unavailable."""
+    pytest.raises(CommandExecutionError, yumpkg.services_need_restart)
```

Comment:
```suggestion
    with pytest.raises(CommandExecutionError):
       yumpkg.services_need_restart()
```
I'm not asking to change here, but you know this also works?

Again, not a required change.